File: LeetCode/DP/JS/Leetcode310-树的最小高度.js
var findMinHeightTrees = function (n, edges) {
    const g = new Array(n).fill(0).map(() => []);
    const d1 = new Array(n).fill(0), d2 = new Array(n).fill(0), p1 = new Array(n).fill(0), up = new Array(n).fill(0);
    for (const ed of edges) {
        a = ed[0], b = ed[1];
        g[a].push(b), g[b].push(a);
    }

    let dfs1 = function (u, fa) {
        for (const x of g[u]) {
            if (x == fa) continue;
            dfs1(x, u);
            const d = d1[x] + 1;
            if (d >= d1[u]) {
                d2[u] = d1[u], d1[u] = d;
                p1[u] = x;
            } else if (d > d2[u]) {
                d2[u] = d;
            }
        }
    }

    let dfs2 = function (u, fa) {
        for (const x of g[u]) {
            if (x == fa) continue;
            if (p1[u] == x) up[x] = Math.max(up[u], d2[u]) + 1;
            else up[x] = Math.max(up[u], d1[u]) + 1;
            dfs2(x, u);
        }
    }

    dfs1(0, -1);
    dfs2(0, -1);

    let mind = n + 1, res = [];
    for (let i = 0; i < n; i++) mind = Math.min(mind, Math.max(up[i], d1[i]));
    for (let i = 0; i < n; i++) {
        if (Math.max(up[i], d1[i]) == mind) {
            res.push(i);
        }
    }
    return res;
};

File: LeetCode/DP/Golang/Leetcode310-树的最小高度.go
package sbgo

import "fmt"

func findMinHeightTrees(n int, edges [][]int) []int {
	g := make([][]int, n)
	d1, d2, up, p1 := make([]int, n), make([]int, n), make([]int, n), make([]int, n)

	for _, ed := range edges {
		a, b := ed[0], ed[1]
		g[a] = append(g[a], b)
		g[b] = append(g[b], a)
	}
	var dfs1 func(u, fa int)
	var dfs2 func(u, fa int)

	dfs1 = func(u, fa int) {
		for _, x := range g[u] {
			if x == fa {
				continue
			}
			dfs1(x, u)
			d := d1[x] + 1
			if d >= d1[u] {
				d2[u], d1[u] = d1[u], d
				p1[u] = x
			} else if d > d2[u] {
				d2[u] = d
			}

		}
	}

	dfs2 = func(u, fa int) {
		for _, x := range g[u] {
			if x == fa {
				continue
			}
			if p1[u] == x {
				up[x] = max(up[u], d2[u]) + 1
			} else {
				up[x] = max(up[u], d1[u]) + 1
			}
			dfs2(x, u)
		}
	}

	dfs1(0, -1)
	dfs2(0, -1)

	mind, res := n+1, []int{}
	for i := 0; i < n; i++ {
		mind = min(mind, max(up[i], d1[i]))
	}

	fmt.Println(mind)

	for i := 0; i < n; i++ {
		if mind == max(up[i], d1[i]) {
			res = append(res, i)
		}
	}

	return res
}
